Adding SELECT FOR JSON Support

anujsarda3 opened the pull request:
### Description

- Implemented SELECT FOR JSON by transforming the query call to built in function tsql_query_to_json
- Created a file forjson.c to implement the function
- Added grammer support for SELECT FOR JSON and for different modes and directives
- This implementation supports SELECT FOR JSON PATH without nested JSON output
- Added Test cases for SELECT FOR JSON PATH without nested examples
- Added the function defination in upgrade scripts and test case in latest schedule

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/57

TASK: BABEL-3669
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

### Issues Resolved

BABEL-3669

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

anujsarda3 committed:
Adding SELECT FOR JSON Support

- Implemented SELECT FOR JSON by transforming the query call to built in function tsql_query_to_json
- Created a file forjson.c to implement the function
- Added grammer support for SELECT FOR JSON and for different modes and directives
- This implementation supports SELECT FOR JSON PATH without nested JSON output
- Added Test cases for SELECT FOR JSON PATH without nested examples
- Added the function defination in upgrade scripts and test case in latest schedule

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Edit in gram-tsql-rule.y and gram-tsql-decl.y

- Changed the position of TSQL_JSON in the definition of unreserved keyword.
- To resolve the conflict

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Added test case for tsql_query_to_json_text

- Test case for tsql_query_to_json_text as run-sql-validation-tests
  check fails

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -11,6 +11,13 @@ RETURNS ntext
 
+-- Helper function to support the FOR JSON clause
+CREATE OR REPLACE FUNCTION sys.tsql_query_to_json_text(query text, mode int, include_null_value boolean,
+           without_array_wrappers boolean,root_name_present boolean, root_name text)
```
Do we need both `root_name_present` and `root_name`? In other words, is there a case where `root_name_present` is true, but `root_name` is null?

anujsarda3 replied:
Yes, it can be possible that the root_name is passed as empty string. As in the TsqlForJSONMakeFuncCall, when converting the input select query into function call tsql_query_to_json_text(), I am passing NULL root_name as empty string, therefore I need a bool variable to check if root_name is present.

anujsarda3 replied:
Have made changes to reduce the number of arguments in the latest commit.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1374,3 +1374,184 @@ returning_list_has_column_name(List *existing_colnames, char *current_colname)
+ * select tsql_query_to_json_text(FORMAT('select a from t where id = %s', @pid) ... )
+ */
+ResTarget *
+TsqlForJSONMakeFuncCall(TSQL_ForClause* forclause, char* src_query, size_t start_location, core_yyscan_t yyscanner)
```
This function and `TsqlForXMLMakeFuncCall()` has something in common. Could you try to extract some common parts and make it as functions so that we can reduce code duplicates?

anujsarda3 replied:
There are some minute changes in every part of the code except where the function call to the FORMAT function is done. For example the name of the directives are different, then handling of "@ in aliases and making function call has extra parameters.

2jungkook replied:
Got it. Then can we extract the FORMAT part?

anujsarda3 replied:
Resolved in latest commit.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1374,3 +1374,184 @@ returning_list_has_column_name(List *existing_colnames, char *current_colname)
+	 * parameter or tsql identifer, starting from left to right in the query string.
+	 */
+	end_param = query;
+	while ((begin_param = strstr(end_param, "\"@")) != NULL)
```
Please add test cases using parameters. Also, please make sure the tests cover both the below `if` and `else if` blocks.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c:
```diff
@@ -1374,3 +1374,184 @@ returning_list_has_column_name(List *existing_colnames, char *current_colname)
+		 * This if block makes sure to consider it as an alias by escaping the content inside
+		 * the square brackets.
+		 */
+		if ((strstr(before_param,"[") != NULL) && ((end_bracket = strstr(begin_param+2,"]")) != NULL))
```
This is not in `TsqlForXMLMakeFuncCall()`. Is this only needed for `FOR JSON` feature?

anujsarda3 replied:
Yes, it is different for SELECT FOR XML and FOR JSON. In FOR JSON it is a valid alias name but FOR XML it is an invalid alias name.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -1436,6 +1436,81 @@ select_no_parens:
 					}
 					$$ = $2;
 				}
+			| select_clause tsql_for_json_clause
```
Please make sure you have test cases for all parser rules you introduced.

2jungkook commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -4251,3 +4329,49 @@ RevokeStmt:
+ * FOR JSON clause can have 3 directives: ROOT, INCLUDE_NULL_VALUES and WITHOUT_ARRAY_WRAPPER.
+ * Map them to ENUM TSQLJSONDirective and String of the ROOT name respectively.
+ */
+tsql_json_common_directive:
```
Is this rule general enough to cover other json-related functions? If not, I would recommend to use a more specific name, e.g., `tsql_for_json_common_directive`.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,163 @@
+		{
+			ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("Column expressions and data sources without names or aliases cannot be formatted as JSON text using FOR JSON clause. Add alias to the unnamed column or table")));
```
Please add a test case for this.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,163 @@
+							 tsql_json_build_object(CStringGetDatum(colname), 
+							 						colval,
+							 						SPI_gettypeid(SPI_tuptable->tupdesc,i),true)->data);
+		}
```
It can be simplified like
```
if (isnull && !include_null_value)
    continue;

appendStringInfo(result,sep);
sep = ",";
appendStringInfo(result, "%s", tsql_json_build_object(..., isnull, ...));
```

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,163 @@
+	appendStringInfoChar(result,'}');
+	if (rownum != SPI_processed-1)
+	{
+		appendStringInfoString(result,",");
```
Please add a test where `(isnull && !include_null_values)` for all attributes and `SPI_processed > 1`.

2jungkook commented on test/JDBC/expected/BABEL-3407-vu-prepare.out:
```diff
@@ -0,0 +1,84 @@
+GO
+
+-- Throws error as ROOT and WITHOUT_ARRAY_WRAPPER cannot be used together
+CREATE VIEW babel_3407_view8 AS
```
`CREATE VIEW` should fail in this case.

anujsarda3 replied:
It is throwing an error. It can be seen in BABEL-3407-vu-verify.out.

2jungkook replied:
It should throw an error for `CREATE VIEW`, not for the select statement. Please check T-SQL behavior.

anujsarda3 replied:
Resolved in latest commit.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	/* If root_name_present is TRUE then WITHOUT_ARRAY_WRAPPER will be FALSE */
+	if(root_name_present)
```
`if (root_name_present)`

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,163 @@
+	switch (mode)
+	{
+		case TSQL_FORJSON_AUTO: /* FOR JSON AUTO */
+			ereport(ERROR,
```
Please add a test case if this code is reachable, otherwise use `assert` or `internal_error`.

anujsarda3 replied:
A test case is already added for SELECT FOR JSON AUTO. Test case name babel_3407_view9 in BABEL-3407-vu-prepare.sql

2jungkook replied:
Are you sure this code is reachable? There is no error message, `AUTO mode is not supported`, in the test file.

anujsarda3 replied:
Have made changes as discussed.

2jungkook commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -2381,10 +2381,16 @@ static void process_select_statement(
+		else if (selectCtx->for_clause()->JSON()) // for JSON
+		{
+			Assert(selectCtx->for_clause()->PATH());
+		}		
```
1. What about `else` case?
2. `Assert(selectCtx->for_clause()->PATH() || selectCtx->for_clause()->AUTO());`?

anujsarda3 replied:
I am not sure about the else case. Are you suggesting to add `Assert(selectCtx->for_clause()->PATH() || selectCtx->for_clause()->AUTO());` in the else case?

anujsarda3 replied:
Resolved in the latest commit

2jungkook commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -11,6 +11,13 @@ RETURNS ntext
+-- Helper function to support the FOR JSON clause
+CREATE OR REPLACE FUNCTION sys.tsql_query_to_json_text(query text, mode int, include_null_value boolean,
+           without_array_wrappers boolean,root_name_present boolean, root_name text)
+RETURNS ntext
```
Why do we use `ntext`?

anujsarda3 replied:
will change it to sys.NVARCHAR

2jungkook replied:
In that case, please also consider a proper typmod, e.g., sys.nvarchar(4000) or sys.nvarchar(max).

anujsarda3 committed:
Updated test cases and edits in code

- Added Test cases for using parameter, all parser rules, errors, include
  null values directive
- Created a function to execute the transformation of query using FORMAT
  function for both FOR JSON and FOR XML
- Edits in name of grammar rules
- Changes in code forjson.c
- Removed unsupported feature handler implementation for SELECT FOR JSON
- Changed return type of function to sys.NVARCHAR

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3407

anujsarda3 committed:
Update in backep_parser file and upgrade script

- Made changes in backend_parser file to reduce the number of arguments
  passed
- Edits in upgrade script

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,148 @@
+{
+	for (int i=0; i< PG_NARGS()-1; i++)
+		if PG_ARGISNULL(i) 
+			PG_RETURN_NULL();
```
We usually use `{}` for multi-line blocks.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -0,0 +1,148 @@
+				(errcode(ERRCODE_DATA_EXCEPTION),
+				 errmsg("invalid query")));
+
+	/* If (root_name_present) is TRUE then WITHOUT_ARRAY_WRAPPER will be FALSE */
```
Please revise the comment as well.

anujsarda3 committed:
Edits in forjson.c

- nit edits in forjson.c

TASK: BABEL-3407
Signed-off-by: Anuj Sarda <sardanuj@amaozn.com>
